File: admin_app/admin_flutter/lib/models/marketplace.dart
class MercadoLivreStatus {
  const MercadoLivreStatus({
    required this.configured,
    required this.connected,
    required this.message,
    this.accountId,
    this.nickname,
    this.siteId,
    this.expiresAt,
    this.lastSyncAt,
  });

  final bool configured;
  final bool connected;
  final String message;
  final String? accountId;
  final String? nickname;
  final String? siteId;
  final DateTime? expiresAt;
  final DateTime? lastSyncAt;

  factory MercadoLivreStatus.fromJson(Map<String, dynamic> json) {
    return MercadoLivreStatus(
      configured: json['configured'] == true,
      connected: json['connected'] == true,
      message: json['message'] as String? ?? '',
      accountId: json['account_id'] as String?,
      nickname: json['nickname'] as String?,
      siteId: json['site_id'] as String?,
      expiresAt: _parseDate(json['expires_at']),
      lastSyncAt: _parseDate(json['last_sync_at']),
    );
  }
}

class MercadoLivreAuthUrl {
  const MercadoLivreAuthUrl({required this.authUrl, required this.state});

  final String authUrl;
  final String state;

  factory MercadoLivreAuthUrl.fromJson(Map<String, dynamic> json) {
    return MercadoLivreAuthUrl(
      authUrl: json['auth_url'] as String? ?? '',
      state: json['state'] as String? ?? '',
    );
  }
}

class MarketplaceProduct {
  const MarketplaceProduct({
    required this.productId,
    required this.name,
    required this.salePrice,
    required this.stockQuantity,
    required this.active,
    required this.listing,
    this.internalCode,
    this.barcode,
  });

  final int productId;
  final String name;
  final String? internalCode;
  final String? barcode;
  final double salePrice;
  final double stockQuantity;
  final bool active;
  final ProductMarketplaceListing listing;

  factory MarketplaceProduct.fromJson(Map<String, dynamic> json) {
    return MarketplaceProduct(
      productId: json['product_id'] as int,
      name: json['name'] as String? ?? '',
      internalCode: json['internal_code'] as String?,
      barcode: json['barcode'] as String?,
      salePrice: _toDouble(json['sale_price']),
      stockQuantity: _toDouble(json['stock_quantity']),
      active: json['active'] != false,
      listing: ProductMarketplaceListing.fromJson(
        json['listing'] as Map<String, dynamic>? ?? const {},
      ),
    );
  }
}

class ProductMarketplaceListing {
  const ProductMarketplaceListing({
    required this.productId,
    required this.provider,
    required this.enabled,
    required this.syncStock,
    required this.syncPrice,
    required this.status,
    required this.condition,
    this.id,
    this.listingId,
    this.title,
    this.categoryId,
    this.listingTypeId,
    this.permalink,
    this.lastError,
    this.lastSyncedAt,
  });

  final int? id;
  final int productId;
  final String provider;
  final String? listingId;
  final bool enabled;
  final bool syncStock;
  final bool syncPrice;
  final String status;
  final String? title;
  final String? categoryId;
  final String? listingTypeId;
  final String condition;
  final String? permalink;
  final String? lastError;
  final DateTime? lastSyncedAt;

  factory ProductMarketplaceListing.fromJson(Map<String, dynamic> json) {
    return ProductMarketplaceListing(
      id: json['id'] as int?,
      productId: json['product_id'] as int? ?? 0,
      provider: json['provider'] as String? ?? 'mercado_livre',
      listingId: json['listing_id'] as String?,
      enabled: json['enabled'] == true,
      syncStock: json['sync_stock'] != false,
      syncPrice: json['sync_price'] != false,
      status: json['status'] as String? ?? 'not_configured',
      title: json['title'] as String?,
      categoryId: json['category_id'] as String?,
      listingTypeId: json['listing_type_id'] as String?,
      condition: json['condition'] as String? ?? 'new',
      permalink: json['permalink'] as String?,
      lastError: json['last_error'] as String?,
      lastSyncedAt: _parseDate(json['last_synced_at']),
    );
  }

  Map<String, dynamic> toUpdateJson({
    bool? enabled,
    bool? syncStock,
    bool? syncPrice,
  }) {
    return {
      'enabled': enabled ?? this.enabled,
      'sync_stock': syncStock ?? this.syncStock,
      'sync_price': syncPrice ?? this.syncPrice,
      'title': title,
      'category_id': categoryId,
      'listing_type_id': listingTypeId,
      'condition': condition,
    };
  }
}

double _toDouble(Object? value) {
  if (value is num) return value.toDouble();
  return double.tryParse(value?.toString() ?? '') ?? 0;
}

DateTime? _parseDate(Object? value) {
  if (value == null) return null;
  return DateTime.tryParse(value.toString());
}

File: backend/app/schemas/marketplace.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class MercadoLivreStatusRead(BaseModel):
    configured: bool
    connected: bool
    message: str
    account_id: str | None = None
    nickname: str | None = None
    site_id: str | None = None
    expires_at: datetime | None = None
    last_sync_at: datetime | None = None
    listing_limit: int | None = None
    enabled_listings: int = 0
    remaining_listings: int | None = None
    pending_jobs: int = 0


class MercadoLivreAuthUrlRead(BaseModel):
    auth_url: str
    state: str


class MercadoLivreAppConfigRead(BaseModel):
    client_id: str | None = None
    redirect_uri: str | None = None
    webhook_url: str | None = None
    configured: bool
    client_secret_configured: bool
    source: str


class MercadoLivreAppConfigUpdate(BaseModel):
    client_id: str = Field(min_length=1, max_length=120)
    client_secret: str | None = Field(default=None, max_length=255)
    redirect_uri: str = Field(min_length=1, max_length=500)
    webhook_url: str | None = Field(default=None, max_length=500)


class ProductMarketplaceListingUpdate(BaseModel):
    enabled: bool
    sync_stock: bool = True
    sync_price: bool = True
    title: str | None = Field(default=None, max_length=180)
    category_id: str | None = Field(default=None, max_length=60)
    listing_type_id: str | None = Field(default=None, max_length=60)
    condition: str = Field(default="new", max_length=20)


class ProductMarketplaceListingRead(ProductMarketplaceListingUpdate):
    id: int | None = None
    product_id: int
    provider: str
    listing_id: str | None = None
    status: str
    permalink: str | None = None
    last_error: str | None = None
    last_synced_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class MarketplaceProductRead(BaseModel):
    product_id: int
    name: str
    internal_code: str | None = None
    barcode: str | None = None
    sale_price: Decimal
    stock_quantity: Decimal
    active: bool
    listing: ProductMarketplaceListingRead


class MercadoLivreListingImportItem(BaseModel):
    listing_id: str
    title: str
    status: str | None = None
    price: Decimal | None = None
    available_quantity: Decimal | None = None
    permalink: str | None = None
    thumbnail: str | None = None
    category_id: str | None = None
    listing_type_id: str | None = None
    condition: str | None = None
    seller_custom_field: str | None = None
    local_product_id: int | None = None
    local_product_name: str | None = None
    already_linked: bool = False


class MercadoLivreImportPreviewRead(BaseModel):
    total: int
    offset: int
    limit: int
    results: list[MercadoLivreListingImportItem]


class MercadoLivreListingLink(BaseModel):
    product_id: int
    listing_id: str = Field(min_length=3, max_length=80)
    enabled: bool = True
    sync_stock: bool = True
    sync_price: bool = True
